working/4200-metaobjects/feature-specification.md

```diff
@@ -0,0 +1,675 @@
+
+For example:
+
+```dart
+class MetaobjectForA implements Type, PrettyPrintable<A> {
+  const MetaobjectForA(); // Every metaobject class must be const-able.
+  String prettyPrint(A a) => A.prettyPrint(a);
+  bool operator ==(Object other) => ...;
+  int get hashCode => ...;
+}
+
+class MetaobjectForB implements Type, PrettyPrintable<B> {
+  const MetaobjectForB(); // Must be const-able.
+  String prettyPrint(B b) => B.prettyPrint(b);
+  bool operator ==(Object other) => ...;
+  int get hashCode => ...;
+}
+```
+
+These classes are implicitly induced by the compiler and analyzer. We
+can't refer to them in user code because the name is a fresh identifier
+such that it doesn't coincide with any name that a developer has written.
+
+However, when a type variable like `X` above is evaluated as an expression,
+the resulting metaobject will be an instance of the class which is the
+metaobject class for the run-time value of `X`.
+
+These classes implement operator `==` and the getter `hashCode`, such that
+they can behave the same as objects of type `Type` do today, when they are
+obtained by evaluating a type literal as an expression.
+
+The static type of the metaobject will be the metaobject class when the
+type literal which is being evaluated as an expression is a compile-time
+constant type (for example, `var myMetaObject = A;`). In the case where the
+type literal is a type variable `X` that has a bound of the form
+`static extends I`, the metaobject has a static type which is a subtype of
+`Type` and a subtype of `I`. This implies that we can use the members of
+`I` on that metaobject. E.g., we can call `X.prettyPrint(x)` in the example
+above.
+
+As a special case (ensuring that this feature does not break existing
+code), the result of evaluating a type literal that denotes a class, a
+mixin, a mixin class, or an enum declaration that does _not_ have a
+`static implements` or `static extends` clause has static type `Type`, and
+it works exactly the same as today. So does the result of evaluating a type
+that isn't introduced by a declaration (that is, a function type, a record
+type, special types like `dynamic`, union types like `T?`  and
+`FutureOr<T>`, etc.)
+
+The previous example showed how we can use metaobjects to provide access to
+static members of a set of classes (or mixins, etc.) without depending on
+the exact class (mixin, etc.) declaration. This basically means that we
+have made the static members _late bound_, because we're calling them via
+normal instance members of the metaobject. In contrast, regular invocations
+of static members and constructors are bound to a specific call target at
+compile time, which means that the call site depends on that declaration.
+
+The next example illustrates how we can use metaobjects to call
+constructors in a similar way (yielding 'late-bound constructors'):
+
+```dart
+abstract class SimpleCreation<X> {
+  X call();
+  X named(int _);
```

Comment:
I haven't read through the rest of the doc nor the conversation, so sorry if this was already suggested, but

Can there be a special case to allow `new()` here? `call()` is pretty niche and rarely used by common Dart code. I think it would be easier to say "this creates a new object" rather than "this lets you use `()` after the identifier, which would be the default constructor if it was a type". It would also make it clear that if you can do

```dart
final x = X();
```

Then you can also do 
```dart
final x = func(X.new);
```

Reply:
Summary: I don't know. `new` looks like a good idea in the case `X.new`, but it doesn't generalize very well.

Reply:
Interesting edge cases. I don't have anything to add to that as you were very thorough, but thanks for taking a look

Reply:
I added this proposal to the document, just for type variables (so `X.new` means `X.call` and `X.new(42)` means `X.call(42)`). I think that makes sense.